milton-processor/index.ts

```diff
@@ -1,40 +1,54 @@
-// Copyright 2018 Google LLC
-//
-// Licensed under the Apache License, Version 2.0 (the 'License');
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an 'AS IS' BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
 import express = require('express');
 import jsdomLib = require('jsdom');
 import domPurify = require('dompurify');
-import Readability = require("readability");
+import Readability = require('readability');
+import ajax = require("xmlhttprequest");
 import fs = require('fs');
 
 const PORT = Number(process.env.PORT) || 8080;
 const app = express();
 const JSDOM = jsdomLib.JSDOM;
 
-const secret = fs.readFileSync('secret', 'utf8');
+const serverSecret = fs.readFileSync('secret', 'utf8').trim();
+function authenticated(secret: string): boolean {
+  return secret == serverSecret
+}
+
+function decipher(htmlString: string): string {
+  const dom = new JSDOM(htmlString).window
+  domPurify(dom).sanitize(htmlString, {WHOLE_DOCUMENT: true, IN_PLACE: true});
+  const readable = new Readability(dom.document).parse();
+  return readable;
+}
 
 app.use(express.json({limit: '5mb'}));
 
 app.post("/extract", (req, res) => {
-  if (req.body.secret !== secret) {
+  if (!authenticated(req.body.secret)) {
     res.status(403).send('invalid secret');
     return;
   }
-  const page = req.body.page;
-  const dom = new JSDOM(page).window
-  domPurify(dom).sanitize(page,  {WHOLE_DOCUMENT: true, IN_PLACE: true});
-  const readable = new Readability(dom.document).parse();
-  res.send(readable);
+
+  res.send(decipher(req.body.page));
+});
+
+app.post("/percolate", (req, res) => {
+  if (!authenticated(req.body.secret)) {
+    res.status(403).send('invalid secret');
+    return;
+  }
+
+  var xhr = new ajax.XMLHttpRequest();
+  xhr.open('GET', req.body.page)
+  xhr.onload = function() {
+    if (xhr.status === 200) {
+      res.send(decipher(xhr.responseText));
+    } else {
+      res.status(xhr.status).send('Unable to fetch page');
```

Comment:
I'm not sure forwarding the response code is correct; if the upstream server returns e.g. 500, milton-processor probably shouldn't forward that to the client cause it would look like the processor itself broke.

I'd be in favor of returning something like 422, and passing the upsteam error code in the error message.

Reply:
yea fair, updated.